storage/schemadb/src/lib.rs

```diff
@@ -259,29 +265,42 @@ impl DB {
 
     /// Writes a group of records wrapped in a [`SchemaBatch`].
     pub fn write_schemas(&self, batch: SchemaBatch) -> Result<()> {
+        let mut put_sizes: HashMap<ColumnFamilyName, usize> = HashMap::new();
+        let mut delete_counts: HashMap<ColumnFamilyName, usize> = HashMap::new();
+
         let db_batch = rocksdb::WriteBatch::new();
-        for (cf_name, key, write_op) in &batch.rows {
+        for (cf_name, rows) in batch.rows {
             let cf_handle = self.get_cf_handle(cf_name)?;
-            match write_op {
-                WriteOp::Value(value) => db_batch.put_cf(cf_handle, &key, &value),
-                WriteOp::Deletion => db_batch.delete_cf(cf_handle, &key),
+            for (key, write_op) in rows {
+                match write_op {
+                    WriteOp::Value(value) => {
+                        let size = put_sizes.entry(&cf_name).or_insert(0);
+                        (*size) += key.len() + value.len();
+
+                        db_batch.put_cf(cf_handle, &key, &value)
+                    }
+                    WriteOp::Deletion => {
+                        let count = delete_counts.entry(cf_name).or_insert(0);
+                        (*count) += 1;
+
+                        db_batch.delete_cf(cf_handle, &key)
+                    }
+                }
+                .map_err(convert_rocksdb_err)?;
             }
-            .map_err(convert_rocksdb_err)?;
         }
 
         self.inner
             .write_opt(&db_batch, &default_write_options())
             .map_err(convert_rocksdb_err)?;
 
-        for (cf_name, key, write_op) in &batch.rows {
-            match write_op {
-                WriteOp::Value(value) => OP_COUNTER.observe(
-                    &format!("db_put_bytes_{}", cf_name),
-                    (key.len() + value.len()) as f64,
-                ),
-                WriteOp::Deletion => OP_COUNTER.inc(&format!("db_delete_{}", cf_name)),
-            };
-        }
+        // Bump counters only after DB write succeeds.
+        put_sizes.into_iter().for_each(|(cf_name, size)| {
+            OP_COUNTER.observe(&format!("db_put_bytes_{}", cf_name), size as f64)
```

Comment:
This is a histogram, right?  If we add the size like this, we are losing sight of the individual key / value pairs, and only accounting for the total size of the batch.  This would result in skewed size observations, leaving us without insight into the true size of writes.

Reply:
Histogram does do quantiles https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile

if that's what you meant..

Reply:
Ugh now I see what you meant. I was being stupid.